main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -0,0 +1,852 @@
+      case -1 => visitor.visitInsn(ICONST_M1)
+      case 0 => if (!isLong) visitor.visitInsn(ICONST_0) else visitor.visitInsn(LCONST_0)
+      case 1 => if (!isLong) visitor.visitInsn(ICONST_1) else visitor.visitInsn(LCONST_1)
+      case 2 => visitor.visitInsn(ICONST_2)
+      case 3 => visitor.visitInsn(ICONST_3)
+      case 4 => visitor.visitInsn(ICONST_4)
+      case 5 => visitor.visitInsn(ICONST_5)
+      case _ if scala.Byte.MinValue <= i && i <= scala.Byte.MaxValue => visitor.visitIntInsn(BIPUSH, i.toInt)
+      case _ if scala.Short.MinValue <= i && i <= scala.Short.MaxValue => visitor.visitIntInsn(SIPUSH, i.toInt)
+      case _ if scala.Int.MinValue <= i && i <= scala.Int.MaxValue => visitor.visitLdcInsn(i.toInt)
+      case _ => visitor.visitLdcInsn(i)
+    }
+    if (isLong && scala.Int.MinValue <= i && i <= scala.Int.MaxValue && i != 0 && i != 1) visitor.visitInsn(I2L)
+  }
+
+  private def compileUnaryExpr(currentClassType: JvmType.Reference,
+                               visitor: MethodVisitor,
+                               jumpLabels: Map[Symbol.LabelSym, Label],
+                               entryPoint: Label)(op: UnaryOperator, e: Expression)(implicit root: Root, flix: Flix): Unit = {
+    // Adding source line number for debugging
+    addSourceLine(visitor, e.loc)
+
+    compileExpression(currentClassType, e, jumpLabels, entryPoint, visitor)
+    op match {
+      case UnaryOperator.LogicalNot =>
+        val condElse = new Label()
+        val condEnd = new Label()
+        visitor.visitJumpInsn(IFNE, condElse)
+        visitor.visitInsn(ICONST_1)
+        visitor.visitJumpInsn(GOTO, condEnd)
+        visitor.visitLabel(condElse)
+        visitor.visitInsn(ICONST_0)
+        visitor.visitLabel(condEnd)
+      case UnaryOperator.Plus => // nop
+      case UnaryOperator.Minus => compileUnaryMinusExpr(visitor)(e.tpe)
+      case UnaryOperator.BitwiseNegate => compileUnaryNegateExpr(visitor)(e.tpe)
+    }
+  }
+
+  /*
+   * For Int8/Int16, we need to truncate and sign extend the result.
+   *
+   * Example:
+   * Suppose we store the value -128 into an Int8 (byte). The number is represented as (in two's complement):
+   *   10000000
+   * But on the JVM, the value is sign extended and stored as an Int32 (int):
+   *   11111111 11111111 11111111 10000000
+   * If we simply negate -128, we get the value 128, which is represented as:
+   *   00000000 00000000 00000000 10000000
+   * But this is greater than the maximum value (127) for an Int8 (byte). We use I2B to convert the Int32 (int) to an
+   * Int8 (byte), which does a truncation and sign extension:
+   *   11111111 11111111 11111111 10000000
+   * And the final value is -128.
+   *
+   * Note that in Java semantics, the unary minus operator returns an Int32 (int), so the programmer must explicitly
+   * cast to an Int8 (byte).
+   */
+  private def compileUnaryMinusExpr(visitor: MethodVisitor)(tpe: Type)(implicit root: Root, flix: Flix): Unit = tpe match {
```

Comment:
I am not sure I understand exactly what this method does, but it should not rely directly on type. However, I can see that it does need *some* extra information.